README.md
# Oriskami node.js bindings 
[![npm version](https://badge.fury.io/js/oriskami.svg)](https://badge.fury.io/js/oriskami)
[![Build Status](https://travis-ci.org/oriskami/oriskami-node.png?branch=master)](https://travis-ci.org/oriskami/oriskami-node)
[![Inline docs](http://inch-ci.org/github/oriskami/oriskami-node.svg?branch=master)](http://inch-ci.org/github/oriskami/oriskami-node)
 

This library helps interact with Oriskami's API, a *risk management intelligence* 
to detect and manage transactions having a risk. Oriskami works by:
1. fusioning your transaction data with external data providers; 
2. applying filters like whitelists, blacklists, rules, or risk scores trained with machine learning; and
3. triggering notifications when a risk is detected (e-mail, slack, signed webhooks, etc.). 

For a quick start see below, otherwise here are few links:
- [**Browse product details**](https://www.oriskami.com/shop)
- [**Sign up**](https://www.oriskami.com/signup)
- [**Get an access token**](https://my.oriskami.com/)
- [**Schedule a demo**](https://www.oriskami.com/shop/schedule-demo)
- [**See the docs**](https://www.oriskami.com/docs)

## Quick Start: Five steps 

1. **Install** the binding:
```js
npm install --save oriskami
```

2. **Setup** your access token:
```js
var Oriskami    = require("oriskami")
  , oriskami    = new Oriskami("YOUR_API_ACCESS_TOKEN", "API_VERSION")
```

3. **Bootstrap** your account with historical data (`EventPast`; *not* analyzed): 
```js
oriskami.EventPast.create({
  "parameters"  :{
    "id": 16646 
  , "email": "abc@gmail.com"
  , "names": "M Abc"
  , "account_creation_time": "2017-05-17 21:50:00"
  , "account_id": "10000"
  , "account_n_fulfilled": "1"
  , "account_total_since_created": "49.40"
  , "account_total_cur": "EUR"
  , "invoice_time": "2017-05-17 21:55:00"
  , "invoice_address_country": "France"
  , "invoice_address_place": "75008 Paris"
  , "invoice_address_street1": "1 Av. des Champs-Élysées"
  , "invoice_name": "M ABC"
  , "invoice_phone1": "0123456789"
  , "invoice_phone2": null
  , "transport_date": "2017-05-18 08:00:00"
  , "transport_type": "Delivery"
  , "transport_mode": "TNT"
  , "transport_weight": "9.000"
  , "transport_unit": "kg"
  , "transport_cur": "EUR"
  , "delivery_address_country": "France"
  , "delivery_address_place": "75008 Paris"
  , "delivery_address_street1": "1 Av. des Champs-Élysées"
  , "delivery_name": "M ABC"
  , "delivery_phone1": "0123450689"
  , "customer_ip_address": "1.2.3.4"
  , "pmeth_origin": "FRA"
  , "pmeth_validity": "0121"
  , "pmeth_brand": "MC"
  , "pmeth_bin": "510000"
  , "pmeth_3ds": "-1"
  , "cart_products": [ " Product ref #12345 ", " Product ref #45678 " ]
  , "cart_details": [{
      "name": "Product ref #12345"
    , "pu": "10.00"
    , "n": "1"
    , "reimbursed": " 0"
    , "available": "1"
    , "amount": "10.00"
    , "cur": "EUR"
    }, {
      "name": "Product ref #45678"
    , "pu": "20.00"
    , "n": "2"
    , "reimbursed": " 0"
    , "available": "1"
    , "amount": "40.00"
    , "cur": "EUR"
    }]
  , "cart_n": "15000"
  , "order_payment_accepted": "2017-05-17 22:00:00"
  , "amount_pmeth": "ABC Payment Service Provider"
  , "amount_discounts": 0.00
  , "amount_products": 20.00
  , "amount_transport": 10.00
  , "amount_total": 30.00
  , "amount_cur": "EUR"
  }

}, (err, res) => /* callback */)
```

4. **Label** events for which the risk (`is_loss`) has materialized: 
```js
oriskami.EventLabel.update({
  "id" : "1"
, "label": "is_loss"
, "value": "false" 
}, (err, res) => /* process results */)
```

5. **Send** live `Events`:
```js
oriskami.events.create({ }, (err, res) => /* process results */)
```
Then you will get notified when a risk is detected.

## Events API: Open Beta (public preview) 

|               | Resource                | C | R | U | D | L     | Test Specs |
|--------------:| ----------------------- |:-:|:-:|:-:|:-:|:-----:|:-------:|
| **Event data**| Event                   | [`{}`](https://oriskami.com/docs/nodejs#create_event)| [`123`](https://oriskami.com/docs/nodejs#retrieve_event) | [`123, {}`](https://oriskami.com/docs/nodejs#update_event) | [`123`](https://oriskami.com/docs/nodejs#delete_event) | [`{}`](https://oriskami.com/docs/nodejs#list_event) | [See on github](https://github.com/oriskami/oriskami-node/blob/master/test/Resources/Event/spec.js) | 
|               | EventPast               | [`{}`](https://oriskami.com/docs/nodejs#create_eventpast)|  |  |  | [`{}`](https://oriskami.com/docs/nodejs#list_eventpast) | [See on github](https://github.com/oriskami/oriskami-node/blob/master/test/Resources/EventPast/spec.js) | 
|               | EventLabel              | | [`123`](https://oriskami.com/docs/nodejs#retrieve_eventlabel) | [`123, {}`](https://oriskami.com/docs/nodejs#update_eventlabel) | [`123`](https://oriskami.com/docs/nodejs#delete_eventlabel) | [`{}`](https://oriskami.com/docs/nodejs#list_eventlabel) | [See on github](https://github.com/oriskami/oriskami-node/blob/master/test/Resources/EventLabel/spec.js)| 
|               | EventLastId             |  |  |  |  | [`{}`](https://oriskami.com/docs/nodejs#list_eventlastid) | [See on github](https://github.com/oriskami/oriskami-node/blob/master/test/Resources/EventLastId/spec.js)| 

+ *C*: Create
+ *R*: Retrieve
+ *U*: Update
+ *D*: Delete
+ *L*: List
+ `123`: resource id 
+ `{}`: JSON with query parameters

## Parsing incoming webhooks

Oriskami signs its webhooks according to the Json Web Token standard [www](https://jwt.io/)—here is 
how to parse incoming requests with express/restify: 

```js
var Oriskami  = require("oriskami")
  , oriskami  = new Oriskami()

oriskami.getPublicKey((err) => {  /* Retrieve public key (async)                */ 
  app.post("/webhook/oriskami"    /* Endpoint defined as a notification         */
  , oriskami.webhook()            /* Middleware that parses the Json Web Tokens */
  , (req, res, next) => {
    console.log(req.json)
  })
}
```

## Code snippets and filter parameters

A few code snippets:

| Action | Code snippet |
|:------------|:-----|
| 1. Change token | `oriskami.set("auth", "your-api-token")` | 
| 2. Add new `resource` | `oriskami.extend("resourceName", "resourcePath")` | 
| 3. Change timeout | `oriskami.set("timeout", 20000) // in ms` | 
| 4. To disable signing of JSON payloads on `nodejs` >= v5 | `oriskami.set("unsigned", true)` | 

Filters available for most resources (`list`).

| Filter        | Default | Example             | Description                   |
| ------------- |:-------:|:--------------------|:------------------------------|
| `limit`       | `10`    | `{"limit":10}`      | At most `10` returned results |
| `gt`          |         | `{"id":{"gt":10}}`  | `id` greater than 10          |
| `gte`         |         | `{"id":{"gte":10}}` | `id` greater than or equal    |
| `lt`          |         | `{"id":{"lt":10}}`  | `id` less than                |
| `lte`         |         | `{"id":{"lte":10}}` | `id` less than or equal       |

## Development

To run the tests, you will need a Oriskami test API key (from your [Oriskami dashboard](https://my.oriskami.com))

```
export ORISKAMI_TEST_TOKEN="your-test-api-key"
npm install -g mocha
npm test
```
*Note: on Windows, use `SET` instead of `export` for setting the `ORISKAMI_TEST_TOKEN` environment variable.*

### [Issues and feature requests](https://github.com/oriskami/oriskami-node/issues)

## Author

- Inspired from [stripe-node](https://github.com/stripe/stripe-node). 
- Developed and maintained by Fabrice Colas for [Oriskami](https://www.oriskami.com). 


CHANGELOG
1.23 June 27, 2018 
 - Fix error when setting status_code to an undefined response payload.
 
1.22 June 25, 2018
 - Also add Json Web Token parsing for responses when they are encoded.
 - Signing of payload with JWT is optional—it's enabled for node >= v5 by default.
 - Enable / disable signing of payloads with query parameter: `?unsigned=true`, and make it stick with `oriskami.set("unsigned", true)`

1.21 June 16, 2018
 - Fix issue with list method; arguments improperly parsed resulting in certain conditions in the callback being called twice.

1.18-20 June 15, 2018 
 - Add async method to retrieve Oriskami's current public key—oriskami.getPublicKey => '/public-key'
 - Add method (oriskami.webhook ) to parse webhooks received from Oriskami—uses Json Web Token + Asymetric signing with public key 
 - Add ability to change host target (for tests): oriskami.set('host', 'another-server.a-domain.com'))
 - Tests: Async-ify and refactor all the test cases and add CRUD(L) tests for EventPast 

1.17 June 12, 2018
 - Remove dependency for lodash.
 - Remove lambda functions from test cases—not suggested for mocha.
 - Remove console logging in test cases.
 - Fix bug related to SSLv3 handshake failure that prevented Travis-CI to run.
 - Add nodejs versions to Travis-CI (v0.10-12, v1-3, v10) and OS (osx).
 - Add loop to test locally the various versions using NVM.

1.13-16.0 January 11, 2018 
 - If in-browser, do not call setTimeout—from http-browserify's http. 
 - Remove unnecessary refs to server-side packages like fs and path. 
 - Help browserify by making all requires statics. 
 - Remove call to 'fs'-related methods that can only run on the server. 

1.12.0 December 29, 2017
 - Switch Ubivar/Oriskami. 

1.11.0 July 19, 2017
 - Add error subtypes for status codes 402, 405, 429 (PaymentRequired,
 MethodNotAllowed, and TooManyRequests), and 500, 502 (InternalServer,
 and BadGateway) 

1.10.0 July 02, 2017
 - Add EventPast 
 - Add functional tests

1.9.0 June 29, 2017
 - Add RouterDataGeocoding, RouterDataGeoip, RouterDataPhone, RouterDataBin, RouterDataBlacklist
 - Add functional tests

1.8.0 June 20, 2017
 - Add RouterDataFeature
 - Add functional test

1.7.2 June 09, 2017
 - Add attribute 'summary' to EventQueue.list() 

1.7.1 June 08, 2017
 - Rename Router* resources

1.7.0 June 08, 2017
 - Add RouterTest

1.6.0 June 08, 2017
 - Add new resource: Router
 - Add functional tests for this resource

1.5.0 June 08, 2017
 - Add new resource: Parameters 
 - Add functional tests for this resource

1.4.2 June 08, 2017
 - Add test spec (FlowBackup), update README.md (Flows, FlowBackup)

1.4.1 June 07, 2017
 - Uri encode ids for retrieve and del resource actions

1.4.0 June 07, 2017
 - Add resource FlowBackup

1.3.0 June 07, 2017
 - Add query-encoded flow_id for retrieve 

1.2.0 June 06, 2017
 - Add new resources: Flows 
 - Add functional tests for these resources

1.1.0 June 06, 2017
 - Add new resources: Filters, Notifiers
 - Add functional tests for these resources

1.0.0 June 05, 2017
 - Add resources: EventNotification, EventLastId, EventLabel, EventQueue,
 EventReview, FilterWhitelist, FilterBlacklist, FilterRulesCustom,
 FilterRulesBase, FilterRulesAI, FilterScoringsDedicated, NotifierEmail,
 NotifierSms, NotifierSlack, NotifierWebhook, NotifierECommerce
 - Add functional tests for each resource
 - Update documentation
 - Fix bug when making requests whose header attributes had a '\n' 

0.8.1-beta September 28, 2015

0.1.15-beta August 3rd, 2015
 - Add support for versioning of API; fix sent headers and test suite.

0.1.14-beta July 21, 2015
 - Add test specs for new generic resources Addresses, ReviewQueues and 
 ReviewerQueueBindings, and add more debugging information to generic specs 

0.1.[11-13]-beta July 20, 2015
 - Add resources ReviewQueues and ReviewerQueueBindings, and update README.md. 
 - Add resource Reviewers and update README.md.

0.1.[5-10]-beta April 28, 2015
 - Handle and propagate error when the socket hangs up
 - Handle and propagate error on sockets and responses
 - Add callback argument to handler of socket errors 
 - Fix missing: self = this, and require of Error

0.1.3-beta17 April 25th, 2015
 - Implement an extend mechanism to add dynamically new resources to the connector 
 - Set of resources is removed from the default.

0.1.3-beta16 April 19th, 2015
 - Add resource routing to keep track of routing status of transactions

0.1.3-beta15 April 10th, 2015
 - Remove status API endpoint and replace it by {uptime, ci} that provide
 uptime status of http endpoints and continuous integration status of the API
 bindings 

0.1.3-beta14 April 9th, 2015
 - Add summary binding to resources; pointing to https://api.ubivar.com/the-resource/summary. 

0.1.3-beta13 April 9th, 2015
 - Add support for multiple inserts [{record #1}, {record #2}, ...]. 

0.1.3-beta12 March 31st, 2015
 - Add support for geoip lookup based on maxmind geoip-lite. 

0.1.3-beta11 March 31st, 2015
 - Add support for bank identification number (BIN).

0.1.3-beta10 March 30th, 2015
 - Enable API-based access to the uptime status of the different web and API
 resources of Ubivar

0.1.3-beta9 March 29th, 2015
 - Add support for historical and daily FX rate from ECB as well as 
 conversion (cur_from, cur_to, date, rate)

0.1.3-beta8 March 28th, 2015
 - Add support for pagination and operators in REST API (gt, gte, lt, lte,
   order, limit, start_after, end_before).
 - Add test cases in ubivar-node client to validate the support of the API 

0.1.3-beta7 March 16th, 2015
 - Fix issue #1 unexpected end of input when creating resource 

0.1.3-beta6 March 9th, 2015
 - Add support for the '/me' route

0.1.3-beta1 March 4th, 2015
 - Initial release with token APIs


lib/Resource/response/index.js
/*
 * Request response handler
 * @req [Object] the submitted request 
 * @next [function] an optional callback
 */
module.exports    = function(req, next) {
  var jwt         = require("jsonwebtoken")
    , https       = require("https")
    , self        = this
    , publicKey   = this.oriskami._api.public_key
    , Error       = require("../../Error")

  return function(res) {
    var body      = ""

    res.setEncoding("utf8")

    res.on("data" , function(chunk) { body += chunk })
    res.on("end"  , function() {
      var err
        , json    = {"success": "false", "data": null, "err": "err_internal"}
        , code    = parseInt(res.statusCode)
        , callback= function(err, json) { 
          json      = json || {}
          json.status_code = code
          next(err, json) 
        }

           if(code === 500) err = new Error.OriskamiInternalServerError(json)
      else if(code === 502) err = new Error.OriskamiBadGatewayError(json)
      else if(code  >  500) err = new Error.OriskamiAPIError(json)
      try { // to parse the json
        json                    = JSON.parse(body)
        if(code === 401 ) err   = new Error.OriskamiUnauthorizedError(json)
        if(code === 402 ) err   = new Error.OriskamiPaymentRequiredError(json)
        if(code === 405 ) err   = new Error.OriskamiMethodNotAllowedError(json)
        if(code === 429 ) err   = new Error.OriskamiTooManyRequestsError(json)
        if(code  <  200 ) err   = new Error.OriskamiError.generate(json)
        if(code  >= 300 && code < 400) err = new Error.OriskamiError.generate(json)
      } catch (e) { // failed to parse the json
        json.err= e.message
        err     = new Error.OriskamiAPIError(json)
      }

      if(json.token && publicKey){ 
        jwt.verify(json.token, publicKey, callback) 
      } else if(json.token && !publicKey){
        var req = https.get("https://api.oriskami.com/public-key", function(res){
          var pkey      = ""
          res.setEncoding("utf8")
          res.on("data", function(chunk){ pkey += chunk })
          res.on("end" , function(){ 
            self.oriskami._api.public_key = pkey 
            jwt.verify(json.token, pkey, callback) 
          })
        }).on("error", callback).end()
      } else {
        callback(err, json)
      }
    })

    res.on("error", function(err){
      err       = new Error.OriskamiAPIError({"success": "false", "data": null, "err": err.message})
      next(err, json)
    })
  }
}
